### PR TITLE
Replace node

### PR DESCRIPTION
**Pull request by snazy**
Now that's #45 is merged, here's the follow-up for "replace node".

To replace a node, it must first be terminated (i.e. executor shutdown).

### CODE REVIEW

File: cassandra-mesos-model/src/main/proto/io/mesosphere/mesos/frameworks/cassandra/model.proto [hunk 1]
```diff
@@ -178,6 +180,8 @@ message CassandraNode {
 
     optional int32 cassandraDaemonPid = 12;
 
+    optional string replacementForIp = 13;
```
Comment (BenWhitehead) [inline cassandra-mesos-model/src/main/proto/io/mesosphere/mesos/frameworks/cassandra/model.proto]: Is this something that will be tracked long term? Do we need to know about the replaced ip after the replacement has taken place?

File: cassandra-scheduler/src/main/java/io/mesosphere/mesos/frameworks/cassandra/ApiController.java [hunk 2]
```diff
@@ -682,6 +695,14 @@ public Response nodeRestart(@PathParam("node") String node) {
         return nodeStatusUpdate(cassandraNode);
     }
 
+    @GET
```
Comment (BenWhitehead) [inline cassandra-scheduler/src/main/java/io/mesosphere/mesos/frameworks/cassandra/ApiController.java]: This should be a `POST`

File: cassandra-scheduler/src/main/java/io/mesosphere/mesos/frameworks/cassandra/ApiController.java [hunk 3]
```diff
@@ -713,4 +734,44 @@ private static Response nodeStatusUpdate(CassandraFrameworkProtos.CassandraNode
         }
         return Response.status(status).entity(sw.toString()).type("application/json").build();
     }
+
+    @GET
```
Comment (BenWhitehead) [inline cassandra-scheduler/src/main/java/io/mesosphere/mesos/frameworks/cassandra/ApiController.java]: This should be a `POST`

File: cassandra-scheduler/src/main/java/io/mesosphere/mesos/frameworks/cassandra/ApiController.java [hunk 4]
```diff
@@ -713,4 +734,44 @@ private static Response nodeStatusUpdate(CassandraFrameworkProtos.CassandraNode
...
+    @Path("/node/replace/{node}")
+    public Response nodeReplace(@PathParam("node") String node) {
```
Comment (BenWhitehead) [inline cassandra-scheduler/src/main/java/io/mesosphere/mesos/frameworks/cassandra/ApiController.java]: What is the use case you are imagining this endpoint satisfies? I thought the scheduler was responsible for making decisions about when a node should be replaced, this seems like we're making it the operators responsibility.

File: cassandra-scheduler/src/main/java/io/mesosphere/mesos/frameworks/cassandra/CassandraCluster.java [hunk 5]
```diff
@@ -306,23 +318,31 @@ public TasksForOffer getTasksForOffer(@NotNull final Protos.Offer offer) {
...
-                        newNode
-                ));
+                CassandraNode newNode = buildCassandraNode(offer, buildSeedNode, replacementForIp);
+                clusterState.nodeAcquired(newNode);
                 node = CassandraNode.newBuilder(newNode);
             } else
                 node = CassandraNode.newBuilder(nodeOption.get());
 
             if (!node.hasCassandraNodeExecutor()) {
+                if (node.getTargetRunState() == CassandraNode.TargetRunState.TERMINATE) {
```
Comment (BenWhitehead) [inline cassandra-scheduler/src/main/java/io/mesosphere/mesos/frameworks/cassandra/CassandraCluster.java]: Do terminated nodes stick around in the state info after we've verified that the node has terminated?  It seems like we're checking the terminated state a lot.

File: cassandra-scheduler/src/main/java/io/mesosphere/mesos/frameworks/cassandra/CassandraCluster.java [hunk 6]
```diff
@@ -977,4 +1039,30 @@ private boolean isLiveNode(CassandraNode node) {
             return false;
         return info.getNativeTransportRunning() && info.getRpcServerRunning();
     }
+
+    public CassandraNode replaceNode(String node) throws ReplaceNodePreconditionFailed {
```
Comment (BenWhitehead) [inline cassandra-scheduler/src/main/java/io/mesosphere/mesos/frameworks/cassandra/CassandraCluster.java]: Please add tests that check both the positive & negative cases of each of these branches. There is a lot of logic here that could go wrong if we don't account for something correctly.

File: cassandra-mesos-model/src/main/proto/io/mesosphere/mesos/frameworks/cassandra/model.proto [hunk 7]
```diff
@@ -97,6 +97,8 @@ message CassandraClusterState {
     repeated CassandraNode nodes = 1;
     repeated ExecutorMetadata executorMetadata = 2;
     optional int64 lastServerLaunchTimestamp = 3;
+    repeated string replaceNodeIps = 4;
+    required int32 nodesToAcquire = 5;
```
Comment (BenWhitehead) [inline cassandra-mesos-model/src/main/proto/io/mesosphere/mesos/frameworks/cassandra/model.proto]: How does this differ from `numberOfNodes` in `CassandraConfigRole`? It seems like the same thing to me.

File: cassandra-scheduler/src/main/java/io/mesosphere/mesos/frameworks/cassandra/CassandraCluster.java (hunk 6 above)
Comment (BenWhitehead) [inline cassandra-scheduler/src/main/java/io/mesosphere/mesos/frameworks/cassandra/CassandraCluster.java]: Code coverage info can be seen here: https://teamcity.mesosphere.io/viewLog.html?buildId=12455&tab=coverage_idea&buildTypeId=Oss_Mesos_Cassandra_CassandraFramework

File: cassandra-mesos-model/src/main/proto/io/mesosphere/mesos/frameworks/cassandra/model.proto (hunk 1 above)
Comment (snazy) [inline cassandra-mesos-model/src/main/proto/io/mesosphere/mesos/frameworks/cassandra/model.proto]: It's not tracked long-term. Once the node has been replaced, the field is cleared.

File: cassandra-scheduler/src/main/java/io/mesosphere/mesos/frameworks/cassandra/ApiController.java (hunk 2 above)
Comment (snazy) [inline cassandra-scheduler/src/main/java/io/mesosphere/mesos/frameworks/cassandra/ApiController.java]: fixed

File: cassandra-scheduler/src/main/java/io/mesosphere/mesos/frameworks/cassandra/ApiController.java (hunk 3 above)
Comment (snazy) [inline cassandra-scheduler/src/main/java/io/mesosphere/mesos/frameworks/cassandra/ApiController.java]: fixed

File: cassandra-scheduler/src/main/java/io/mesosphere/mesos/frameworks/cassandra/ApiController.java (hunk 4 above)
Comment (snazy) [inline cassandra-scheduler/src/main/java/io/mesosphere/mesos/frameworks/cassandra/ApiController.java]: Nope - the operator's responsible to replace a node. It's nothing that can or should be done automatically.

Of course, there are multiple scenarios:
* manual, planned node replacement: for example when you scale up one node after another (replacing with "better" hardware)
* determining that a node is no longer available (specific to Cassandra-on-Mesos): But when exactly is a node "no longer available"?

We should be very careful about doing such things automatically.

File: cassandra-scheduler/src/main/java/io/mesosphere/mesos/frameworks/cassandra/CassandraCluster.java (hunk 5 above)
Comment (snazy) [inline cassandra-scheduler/src/main/java/io/mesosphere/mesos/frameworks/cassandra/CassandraCluster.java]: Currently, yes. We should "lock out" the terminated and replaced node for some days until the last node in the cluster has forgotten the replaced node. (It may take some time until the last node recognizes that - such a "last" node could be one that has been dead for some time, too.)

Sure, we re-check such stuff as soon as we have permanently allocated resources from Mesos (get the same data volumes and ports from Mesos after a restart of the executor)

File: cassandra-mesos-model/src/main/proto/io/mesosphere/mesos/frameworks/cassandra/model.proto (hunk 7 above)
Comment (snazy) [inline cassandra-mesos-model/src/main/proto/io/mesosphere/mesos/frameworks/cassandra/model.proto]: That's the "current" number of new nodes (executors) to acquire. It's decremented once an executor is acquired.
Initially it's equal to `numberOfNodes`. But once you call `replaceNode` is incremented.

File: cassandra-scheduler/src/main/java/io/mesosphere/mesos/frameworks/cassandra/CassandraCluster.java (hunk 6 above)
Comment (snazy) [inline cassandra-scheduler/src/main/java/io/mesosphere/mesos/frameworks/cassandra/CassandraCluster.java]: k, added more tests
